multi-architecture builds and publish

jon-nfc opened the pull request:
### SUMMARY

Add changes related to multi-architecture builds and publishing. 

The desired endstate is that there is officially built and supported multi-architecture container images that can be built from a single command `make docker-buildx` and that on quay.io/ansible/awx-operator those same muli-architecture images are available for consumption.


#### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New or Enhanced Feature


### ADDITIONAL INFORMATION

- Current release runbook can be found in [docs/contributors-guide/release-process.md](https://github.com/ansible/awx-operator/blob/07427be0b79411f5c8fd2ffd7caf4fd92fcce47a/docs/contributors-guide/release-process.md)

This PR does not have the intention of changing the release workflow. The intent is that as part of the established workflow, that multi-architecture containers are built and published for use by the community.

Adjustment was made to ensure that there is ever only one image built for a given git commit and is used for the entire workflow. This prevents the scenario where it's as if a witch has casted a spell upon you, when at the end of the day there were two or more images that exist for the same commit and you were the poor sole who can't figure out what is wrong because the image you have is "Good to go!!" Bottom line no one wants to debug black magic (be in a scenario where two images exist for the same code but are different enough to make you pull your hair out). As each commit is stored on GHCR, it becomes the single source of truth for any containers built, especially for debugging purposes.

#### Changes to Github Actions

- **devel** _push to devel branch_  builds multi-arch containers and tags it with the git commit and stores it on GHCR in addition to the existing step of pushing said manifest to quay.io with the `devel` tag

- **stage** _when triggered_ no longer builds it's own image. pulls the image from GHCR built in `devel` action and follows the existing flow of tagging with the specified version and pushing back to GHCR

- **promote** _Github Release_ No change, however now uses the multi-arch manifest.

#### Other Changes

reference: ansible/eda-server-operator#158

- makefile now has a target `docker-buildx` this target provides the basis for all multi-arch container builds. As buildx creates a manifest with all container architectures included, as part of the same command will automagic push the image to the tagged registry. This command can also be used in development 

    - from a terminal `make docker-buildx`
    - To customize the image name and registry `IMG=registry/name/name2:tag make docker-buildx`. 
    - To specifiy different architectures to build `PLATFORMS=linux/arm64,linux/amd64 make docker-buildx`
    - multiple variables can be addded to the command should it be desired.

### Fixes issues

- #31
- #1680

jon-nfc committed:
Add multi-arch image builds to makefile

A default set of platforms is specified and will build those if var `PLATFORMS` is not specified on the CLI

Issue #31 fixes #1680

jon-nfc committed:
feat(workflow): update devel workflow to build multi-arch containers.

1. builds containers and stores @ghcr with tag that is git commit.
2. pushes the image to quay.io/ansible with tag devel.

Issue #31

jon-nfc committed:
feat(workflow): Update stage to stage image from single build point

It's not good practice to rebuild images as the scenario exists where the same code base builds different images. As such only use images from a single build point.

1. Pulls the image that matches the git commit
2. Tag image with the specified version and push to ghcr

Issue #31

jon-nfc committed:
feat(workflow): Update Promote to use single build point and the multi-arch manifest

1. Pulls the complete manifest containing all architectures that matches the version from ghcr
2. Tag manifest with the specified version and publish to quay.io
3. Tag manifest as 'latest' and publish to quay.io

fixes #31

jon-nfc commented:
https://github.com/ansible/awx-operator/blob/07427be0b79411f5c8fd2ffd7caf4fd92fcce47a/.github/workflows/publish-operator-hub.yaml#L78-L86

which leads to

https://github.com/ansible/awx-operator/blob/07427be0b79411f5c8fd2ffd7caf4fd92fcce47a/hack/publish-to-operator-hub.sh#L43-L45

As part of the publish there is another build and push of the image. Can this be safely removed? AS the complete manifest is already pushed to quay.io in the `promote` action

https://github.com/ansible/awx-operator/blob/07427be0b79411f5c8fd2ffd7caf4fd92fcce47a/Makefile#L235-L237

Also I noticed that step `make catalog-build` uses something called opm, what is it and in it's current usage, can it handle docker manifests?

jon-nfc commented:
Don't know who to ping, PR ready for review.

also require someone in the know to answer my above question

rooftopcellist commented:
@jon-nfc 

> As part of the publish there is another build and push of the image. Can this be safely removed? AS the complete manifest is already pushed to quay.io in the promote action

The `publish-to-operator-hub.sh` script is for adding the awx-operator version to https://operatorhub.io/operator/awx-operator so it can be installed with OLM.  The bundle image is needed for OLM installs, it contains content from the `manifests` and `crd` directories (critically, the CSV). 

* **bundle image build** - This image is separate from the operator image. It is necessary for OLM installs, but shouldn't be any trouble for multi-arch.

> Bundle images are not architecture-specific. They contain only plaintext Kubernetes manifests and operator metadata. (Reference)[https://sdk.operatorframework.io/docs/advanced-topics/multi-arch/#operator-lifecycle-manager]


* **catalog image build** - not strictly necessary, but it is a nice-to-have because it allows for easier testing of OLM-specific parameter changes. Also handy for demos of new features because there can sometimes be a 1-2 day lag for the operator getting published to OperatorHub.io after a release.  

----------------------------------------------------

The `opm` tool is a cli tool for inspecting and managing catalog images.  More information can be found [here](https://github.com/operator-framework/operator-registry/blob/master/docs/design/opm-tooling.md).

A catalog image is a 3rd image that contains pointers to operator bundle versions it knows about, which populates the list of available operators that can be installed on Openshift or an OLM enabled k8s cluster.  The newer catalog images are just plaintext on the inside as far as I know (file-based-catalogs) as far as I know.  

This is not strictly needed for users as the folks from https://github.com/k8s-operatorhub/community-operators maintain a catalog image that gets installed as part of OLM when you install an operator from OperatorHub.io.  So I don't think we need to worry about multi-arch for these images.

rooftopcellist commented:
@jon-nfc it looks like there are a few linting issues that need to be cleaned up. 

```
./.github/workflows/promote.yaml
  Warning: 32:76 [comments] too few spaces before comment
  Warning: 40:76 [comments] too few spaces before comment
  Error: 49:171 [line-length] line too long (176 > 170 characters)
  Error: 54:171 [line-length] line too long (171 > 170 characters)

./.github/workflows/stage.yml
  Warning: 58:76 [comments] too few spaces before comment
  Error: 67:171 [line-length] line too long (216 > 170 characters)

./.github/workflows/devel.yaml
  Warning: 17:76 [comments] too few spaces before comment
  Warning: 25:76 [comments] too few spaces before comment
  Error: [39](https://github.com/ansible/awx-operator/actions/runs/7447379325/job/20266238173?pr=1681#step:6:43):171 [line-length] line too long (176 > 170 characters)
  Error: [40](https://github.com/ansible/awx-operator/actions/runs/7447379325/job/20266238173?pr=1681#step:6:44):1 [empty-lines] too many blank lines (1 > 0)
```

rooftopcellist commented on .github/workflows/stage.yml:
```diff
@@ -53,17 +53,19 @@ jobs:
+
+      - name: Stage awx-operator
         run: |
-          BUILD_ARGS="--build-arg DEFAULT_AWX_VERSION=${{ github.event.inputs.default_awx_version }} \
```
@jon-nfc The `DEFAULT_AWX_VERSION` and `OPERATOR_VERSION` build-args still need to be passed here as they are used by the Dockerfile to set env vars and are consequently used in the operator's roles.

rooftopcellist replied:
@jon-nfc Unfortunately, since we need to be able to pin the version of the awx image at build time of the operator, that kind poses a problem for this approach of using the pre-built images from `devel`, re-tagging, and releasing them...

jon-nfc replied:
the origin of the "known" image in the chain I think can still be maintained by keeping the dev build as that, dev/testing; however independent of the release cycle. The release image chain could be started from the `stage` action and then remain the same as it appears this is where the version number is added within the workflow.

#### Current HEAD

https://github.com/ansible/awx-operator/blob/07427be0b79411f5c8fd2ffd7caf4fd92fcce47a/Dockerfile#L1-L25

https://github.com/ansible/awx-operator/blob/07427be0b79411f5c8fd2ffd7caf4fd92fcce47a/.github/workflows/devel.yaml#L1-L27

https://github.com/ansible/awx-operator/blob/07427be0b79411f5c8fd2ffd7caf4fd92fcce47a/.github/workflows/stage.yml#L1-L85

jon-nfc replied:
see 78317ff54e21c3e86ed57846907d36dbd8a10626 and 1a9404954d52485cae583080c19b18719fbcd1d9 for the implementation.

these changes enable the building of the image from the stage action, which sets the container ENV vars as per the original command.

jon-nfc commented on .github/workflows/stage.yml:
```diff
@@ -53,17 +53,19 @@ jobs:
-          BUILD_ARGS="--build-arg DEFAULT_AWX_VERSION=${{ github.event.inputs.default_awx_version }} \
-                      --build-arg OPERATOR_VERSION=${{ github.event.inputs.version }}" \
-          IMAGE_TAG_BASE=ghcr.io/${{ github.repository_owner }}/awx-operator \
-          VERSION=${{ github.event.inputs.version }} make docker-build docker-push
```
`version` Although specified in the original build command, I can't find where this is used. It's is referenced in the makfile, however for the original docker-builld/push not used. Or have i missed something?

edit add what talking about: version

rooftopcellist replied:
>VERSION=${{ github.event.inputs.version }} make docker-build docker-push

`VERSION` is used when building the `IMG` variable, which is used when tagging and pushing the image. That happens here:
* https://github.com/ansible/awx-operator/blob/devel/Makefile#L60

rooftopcellist replied:
What you did instead of setting `IMG` directly is a bit clearly, and I think is better.

jon-nfc committed:
fix(makefile): build args required for the container build

As there are ENV vars set by the build process, there needs to be a way of passing them at build time

PR #1681

jon-nfc committed:
fix(workflow): container requires that ENV vars be set.

As part of the release process, env vars for the operator and awx must be set. As such the image needs to be rebuilt.

PR #1681

jon-nfc committed:
chore: correct liniting errorr

PR #1681

jon-nfc commented:
> This is not strictly needed for users as the folks from https://github.com/k8s-operatorhub/community-operators maintain a catalog image that gets installed as part of OLM when you install an operator from OperatorHub.io. So I don't think we need to worry about multi-arch for these images.

I though as much.

rooftopcellist commented:
@jon-nfc I think this is a good approach.  I didn't spot any issues while reading this through.  I would like to run through a mock release on my fork first though before merging.  I'll plan on doing that tomorrow.

cc @TheRealHaoLiu for review as well.

jon-nfc committed:
chore: correct linting error

42:1 [empty-lines] too many blank lines (1 > 0)

PR #1681

jon-nfc commented:
> @jon-nfc I think this is a good approach. I didn't spot any issues while reading this through except for one more linting error CI is complaining about. I would like to run through a mock release on my fork first though before merging. I'll plan on doing that tomorrow.
> 
> ```
> ./.github/workflows/devel.yaml
>   Error: 42:1 [empty-lines] too many blank lines (1 > 0)
> ```
> 
> cc @TheRealHaoLiu for review as well.

Righto, I'll have everything good to go in the next few hours.

> tomorrow

depends on where you are, where i am it could still be yesterday for you.

rooftopcellist commented on .github/workflows/stage.yml:
```diff
@@ -53,17 +53,19 @@ jobs:
-          IMAGE_TAG_BASE=ghcr.io/${{ github.repository_owner }}/awx-operator \
-          VERSION=${{ github.event.inputs.version }} make docker-build docker-push
+              --build-arg OPERATOR_VERSION=${{ github.event.inputs.version }}" \
+          IMG=ghcr.io/${{ github.repository_owner }}/${{ github.repository }}:${{ github.event.inputs.version }} \
```
```suggestion
          IMG=ghcr.io/${{ github.repository }}:${{ github.event.inputs.version }} \
```

We need to remove `${{ github.repository_owner }}` here because it is already there from when it is set during the checkout:
* https://github.com/ansible/awx-operator/pull/1681/files#diff-ca6f66a88a8c14e7412271d092d23b124f23052eb38134788e5a992f0522f574R49

jon-nfc replied:
https://github.com/ansible/awx-operator/blob/4d7936515c89ab9c0be6fb985fd968640874bf26/.github/workflows/stage.yml#L46-L50

I did not know this, however note taken.

question, re variable preceedence

https://github.com/ansible/awx-operator/blob/4d7936515c89ab9c0be6fb985fd968640874bf26/.github/workflows/stage.yml#L40-L50

there are two checkouts, which variable takes precedence? first set, last set?

rooftopcellist commented on .github/workflows/devel.yaml:
```diff
@@ -13,15 +13,29 @@ jobs:
+        run: |
+          docker buildx imagetools create \
+            ghcr.io/${{ github.repository_owner }}/${{ github.repository }}:${{ github.sha }} \
+            --tag quay.io/ansible/${{ github.repository }}:devel
```
```suggestion
            --tag quay.io/ansible/${{ github.repository }}:devel

```

To satisfy the linter check

jon-nfc replied:
done in 4d7936515c89ab9c0be6fb985fd968640874bf26

jon-nfc committed:
chore: correct linting error

molecule (--skip-tags=replicas): .github/workflows/devel.yaml#L41
41:65 [new-line-at-end-of-file] no new line character at the end of file

PR #1681

jon-nfc commented on .github/workflows/devel.yaml:
```diff
@@ -13,15 +13,29 @@ jobs:
+        run: |
+          docker buildx imagetools create \
+            ghcr.io/${{ github.repository_owner }}/${{ github.repository }}:${{ github.sha }} \
+            --tag quay.io/ansible/${{ github.repository }}:devel
```
Don't know if there is an issue with the linter but it inconsistent.

- 466b0cf5c784894e3a0139e5c9014401c52452f5 too many blank lines `0>1` when there were two blank lines https://github.com/ansible/awx-operator/actions/runs/7457722546
- 4d7936515c89ab9c0be6fb985fd968640874bf26 no blank line at end of file `no new line character at the end of file` https://github.com/ansible/awx-operator/actions/runs/7470471612

jon-nfc commented:
branch updated with rebase.

PR ready for review

TheRealHaoLiu commented on .github/workflows/promote.yaml:
```diff
@@ -27,21 +27,36 @@ jobs:
         run: |
-          echo ${{ secrets.QUAY_TOKEN }} | docker login quay.io -u ${{ secrets.QUAY_USER }} --password-stdin
+          docker buildx imagetools create \
+            ghcr.io/${{ github.repository_owner }}/${{ github.repository }}:${TAG_NAME} \
+            --tag quay.io/ansible/${{ github.repository }}:${TAG_NAME}
```
based on https://docs.github.com/en/enterprise-cloud@latest/actions/learn-github-actions/variables#default-environment-variables

GITHUB_REPOSITORY | The owner and repository name. For example, octocat/Hello-World.

i think we need to take in a variable for "which quay repo to push this to"

jon-nfc replied:
@TheRealHaoLiu 

i have an outstanding question above re this from the review that @rooftopcellist did.  That's why the discussions are not resolved as the next reviewer can see and resolve as completed and as appropriate.

https://github.com/ansible/awx-operator/pull/1681#discussion_r1448499326. this question still remains!!!!

TheRealHaoLiu commented on .github/workflows/promote.yaml:
```diff
@@ -27,21 +27,36 @@ jobs:
-          docker tag ghcr.io/${{ github.repository }}:${TAG_NAME} quay.io/${{ github.repository }}:latest
-          docker push quay.io/${{ github.repository }}:${TAG_NAME}
-          docker push quay.io/${{ github.repository }}:latest
+          docker buildx imagetools create \
+            ghcr.io/${{ github.repository_owner }}/${{ github.repository }}:${TAG_NAME} \
+            --tag quay.io/ansible/${{ github.repository }}:latest
```
same as previous comment

TheRealHaoLiu commented on .github/workflows/devel.yaml:
```diff
@@ -13,15 +13,29 @@ jobs:
+
+      - name: Build and Store Image @ghcr
+        run: |
+          IMG=ghcr.io/${{ github.repository_owner }}/${{ github.repository }}:${{ github.sha }} make docker-buildx
```
```suggestion
          IMG=ghcr.io/${{ github.repository }}:${{ github.sha }} make docker-buildx
```

rooftopcellist committed:
Modify Release GHA to run on fork

rooftopcellist commented:
@jon-nfc ,

@TheRealHaoLiu and I tested out this workflow on our fork today and got it working with the changes in this PR.

Mainly, we modified how it runs so that we could test it on a fork and push the images to our personal quay registry so as not to affect the official awx-operator images while testing.

There were also a few variable refs that needed to be changed for the workflow to work. Notably:

* ghcr.io/${{ github.repository_owner }}/${{ github.repository }} evaluates to ghcr.io/ansible/ansible/awx-operator, which failed.

I also changed how it creates the draft release so that it uses the github action for it rather than custom logic to git the GitHub API.

I opened a PR to your branch here:
* https://github.com/jon-nfc/awx-operator/pull/1

Note that a variable needs to be created in the GitHub repo if you want to override the default when triggering the promote pipeline by publishing the draft release:
![image](https://github.com/ansible/awx-operator/assets/11698892/e3cb3130-478d-4b9e-893d-80b9d5870781)


Here is the successful test release on my fork:
* https://github.com/rooftopcellist/awx-operator/actions/runs/7546887214
* https://github.com/rooftopcellist/awx-operator/actions/runs/7547409029

jon-nfc committed:
Merge pull request #1 from rooftopcellist/multi-arch-builds-test

Modify Release GHA to run on fork
